python/ray/data/iterator.py

```diff
@@ -56,6 +59,285 @@ def __iter__(self):
         return self.iterator_gen()
 
 
+DataBatchType = TypeVar("DataBatchType", bound=DataBatch)
+
+
+@DeveloperAPI
+class CollateFn(Generic[DataBatchType]):
+    """A function that converts a DataBatch to a CollatedData."""
+
+    def __init__(
+        self,
+        dtypes: Optional[Union["torch.dtype", Dict[str, "torch.dtype"]]] = None,
+        device: Optional[str] = None,
+    ):
+        """Initialize the collate function.
+
+        Args:
+            dtypes: Optional torch dtype(s) for the tensors
+            device: Optional device to place tensors on
+        """
+        self.dtypes = dtypes
+        self.device = device
+
+    @abc.abstractmethod
+    def __call__(self, batch: DataBatchType) -> "CollatedData":
+        """Convert a batch of data to collated format.
+
+        Args:
+            batch: The input batch to collate
+
+        Returns:
+            The collated data in the format expected by the model
+        """
+        ...
+
+
+@DeveloperAPI
+class ArrowBatchCollateFn(CollateFn["pyarrow.Table"]):
+    """Collate function for converting Arrow tables to PyTorch tensors."""
```

Comment:
`Collate function that takes pyarrow.Table as the input batch type.`
note, this doesn't necessarily have to produce pytorch tensors.